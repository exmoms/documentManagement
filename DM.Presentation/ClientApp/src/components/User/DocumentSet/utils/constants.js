const FILE = 0;
const FOLDER = 1;

export { FILE, FOLDER };
