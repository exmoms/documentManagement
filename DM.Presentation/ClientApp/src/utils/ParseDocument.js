/**
 * This function is responsible of parsing document data (from api/document/getDocumentVersionById?id={id}) into a form readable
 * By react-jsonschema-form library
 * @param {Object}  res : the document as json object
 * @param {Number} option : its value takes 0 for preview document, otherwise for update document
 * @return {Object} consist of four parts:
 *      1. schema: the schema of the form expected by react-jsonschema-form library
 *      2. ui_schema: the wedgit corresponding to each form field.
 *      3. form_data: the form fields data
 *      3. map:  maping attributeName and corresponding attributeId
 */

function parseDocumentDataToForm(res, option) {
  let schema = {
    title: "",
    type: "object",
    required: [],
    properties: {
      metadataModelName: { type: "string", title: "Metadata Model Name" },
      addedDate: { type: "string", title: " Document Added Date" },
      latestVersion: { type: "number", title: "Latest Version" },
      versionMessage: { type: "string", title: "Version Message" },
      documentVersionAddedDate: {
        type: "string",
        title: "Document Version Added Date",
      },
    },
  };

  let form_data = {
    metadataModelName: res.metadataModelName,
    addedDate: res.addedDate.substr(0, 10),
    latestVersion: res.latestVersion,
    versionMessage: res.documentVersion.versionMessage,
    documentVersionAddedDate: res.documentVersion.addedDate.substr(0, 10),
  };

  let ui_schema = {
    metadataModelName: {
      "ui:readonly": true,
    },
    addedDate: {
      "ui:readonly": true,
    },
    latestVersion: {
      "ui:readonly": true,
    },
    versionMessage: {
      "ui:readonly": true,
    },
    documentVersionAddedDate: {
      "ui:readonly": true,
    },
  };

  let map = [];
  schema.title = res.name;

  // if option is update we provide a required "Document Version" field.
  if (option === 1) {
    schema.required.push("documentVersion");
    Object.defineProperty(schema.properties, "documentVersion", {
      value: {
        type: "string",
        title: "New Version Message",
      },
      writable: true,
      enumerable: true,
      configurable: true,
    });
  }
  // loop over all metadata attributes in the documents
  res.documentVersion.values.forEach((element) => {
    map[element.attributeName] = element.attributeId;
    if (element.isRequired) {
      schema.required.push(element.attributeName);
    }
    const name = element.attributeName;

    // if the option = 0 then we should add readonly property to the ui_schema for this field..
    if (option === 0) {
      Object.defineProperty(ui_schema, name, {
        value: { "ui:readonly": true },
        writable: true,
        enumerable: true,
        configurable: true,
      });
    }

    Object.defineProperty(schema.properties, name, {
      value: {
        type: "",
        title: name,
      },
      writable: true,
      enumerable: true,
      configurable: true,
    });

    switch (element.typeId) {
      case 1:
        schema.properties[name].type = "boolean";
        break;
      case 2:
        schema.properties[name].type = "string";
        if (ui_schema.hasOwnProperty(name)) {
          Object.defineProperty(ui_schema, name, {
            value: {
              "ui:widget": "date",
              "ui:readonly": true,
            },
            writable: true,
            enumerable: true,
            configurable: true,
          });
        } else {
          Object.defineProperty(ui_schema, name, {
            value: {
              "ui:widget": "date",
            },
            writable: true,
            enumerable: true,
            configurable: true,
          });
        }

        break;
      case 3:
      case 4:
      case 5:
        schema.properties[name].type = "number";
        break;
      case 6:
        schema.properties[name].type = "string";
        break;
      default:
      //@ todo: we should return error code to indecate unsupported types.
    }

    Object.defineProperty(form_data, name, {
      value: element.typeId !== 2 ? element.value : element.value.substr(0, 10),
      writable: true,
      enumerable: true,
      configurable: true,
    });
  });

  return { schema, ui_schema, form_data, map };
}

export default parseDocumentDataToForm;
