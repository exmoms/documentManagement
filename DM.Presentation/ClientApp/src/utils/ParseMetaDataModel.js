/**
 * This function is responsible of parsing metadata model data (from api/document/getDocumentVersionById?id={id}) into a form readable
 * By react-jsonschema-form library
 * @param {Object}  res : the metadata Model as json form
 * @return {Object} consist of three parts:
 *      1. json_schema: the schema of the form expected by react-jsonschema-form library
 *      2. ui_schema: the wedgit corresponding to each form field.
 *      3. map:  maping metaDataAttributeName and corresponding metaDataAttributeId
 */

function parseMetaDataModel(res) {
  let json_schema = {
    title: "",
    type: "object",
    required: [],
    properties: {},
  };

  let ui_schema = {};

  let map = [];
  json_schema.title = res.metaDataModelName;
  res.metaDataAttributes.forEach((element) => {
    map[element.metaDataAttributeName] = element.id;
    if (element.isRequired) {
      json_schema.required.push(element.metaDataAttributeName);
    }
    const name = element.metaDataAttributeName;
    Object.defineProperty(json_schema.properties, name, {
      value: {
        type: "",
        title: "",
        default: "",
      },
      writable: true,
      enumerable: true,
      configurable: true,
    });

    json_schema.properties[name].title = name;
    const type = element.dataTypeID;
    if (type === 6) {
      json_schema.properties[name].type = "string";
    } else if (type === 3 || type === 5 || type === 4) {
      json_schema.properties[name].type = "number";
    } else if (type === 1) {
      json_schema.properties[name].type = "boolean";
      json_schema.properties[name].default = false;
    } else if (type === 2) {
      json_schema.properties[name].type = "string";
      Object.defineProperty(ui_schema, name, {
        value: {
          "ui:widget": "date",
        },
        writable: true,
        enumerable: true,
        configurable: true,
      });
    } else {
      //@ todo: we should return error code to indecate unsupported types.
    }
  });
  return { json_schema, ui_schema, map };
}

export default parseMetaDataModel;
