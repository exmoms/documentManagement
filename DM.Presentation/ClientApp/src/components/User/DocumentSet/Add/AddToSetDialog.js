/**
 * @file Contains AddToSetDialog component implementaion
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @version 1.0
 */

import React, { Fragment } from "react";
import { makeStyles } from "@material-ui/core/styles";
import Avatar from "@material-ui/core/Avatar";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import ListItemText from "@material-ui/core/ListItemText";
import DialogTitle from "@material-ui/core/DialogTitle";
import Dialog from "@material-ui/core/Dialog";
import AddIcon from "@material-ui/icons/Add";
import { blue } from "@material-ui/core/colors";
import { FOLDER, FILE } from "../utils/constants";
import { useTranslation } from "react-i18next";

const useStyles = makeStyles({
  avatar: {
    backgroundColor: blue[100],
    color: blue[600],
  },
});

export default function AddToSetDialog(props) {
  const classes = useStyles();
  const { show, set } = props;

  const handleClose = () => {
    props.action();
  };

  const { t } = useTranslation();
  let lang = localStorage.getItem("i18nextLng");
  return (
    <Dialog
      onClose={handleClose}
      aria-labelledby="simple-dialog-title"
      open={show}
      maxWidth={"xs"}
    >
      <div
        style={lang === "ar" ? { textAlign: "right" } : { textAlign: "left" }}
      >
        <DialogTitle id="simple-dialog-title">{t("actions")}</DialogTitle>
      </div>
      <List>
        <ListItem autoFocus button onClick={() => props.action(FOLDER, true)}>
          <ListItemAvatar>
            <Avatar className={classes.avatar}>
              <AddIcon />
            </Avatar>
          </ListItemAvatar>
          <ListItemText primary={t("documentset_addnewdocumentset")} />
        </ListItem>

        {set.id !== 0 ? (
          <Fragment>
            <ListItem
              autoFocus
              button
              onClick={() => props.action(FOLDER, false)}
            >
              <ListItemAvatar>
                <Avatar className={classes.avatar}>
                  <AddIcon />
                </Avatar>
              </ListItemAvatar>
              <ListItemText primary={t("add_existing_set")} />
            </ListItem>
            <ListItem autoFocus button onClick={() => props.action(FILE)}>
              <ListItemAvatar>
                <Avatar className={classes.avatar}>
                  <AddIcon />
                </Avatar>
              </ListItemAvatar>
              <ListItemText primary={t("add_document_to_set")} />
            </ListItem>
          </Fragment>
        ) : (
          ""
        )}
      </List>
    </Dialog>
  );
}
