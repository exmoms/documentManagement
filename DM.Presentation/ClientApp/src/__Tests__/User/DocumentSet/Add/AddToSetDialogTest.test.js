import React from "react";
import ReactDom from "react-dom";
import AddToSetDialog from "../../../../components/User/DocumentSet/Add/AddToSetDialog";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { mount } from "enzyme";
import { createMount } from '@material-ui/core/test-utils';
import Dialog from "@material-ui/core/Dialog";
import { Fragment } from "react";
import ListItem from "@material-ui/core/ListItem";
import { FOLDER, FILE } from "../../../../components/User/DocumentSet/utils/constants";
afterEach(cleanup);

describe("AddToSetDialog", () => {

    test("renders without crashing", () => {
        let show = true;
        let set={id:1};
        const action = jest.fn();
        const div = document.createElement("div");
        ReactDom.render(<AddToSetDialog show={show} set={set} action={action}/>, div);
      });

    test("AddToSetDialog matches snapshot", () => {
        let show = true;
        let set={id:1};
        const action = jest.fn();
    const renderedValue = createMount()(
        <AddToSetDialog show={show} set={set} action={action}/>
      );
      expect(renderedValue.html()).toMatchSnapshot();
    });
   
    test("AddToSetDialog test intialization", () => {
        let show = true;
        let set1={id:0};
        const action = jest.fn();
        const wrapper = mount(
            <AddToSetDialog show={show} set={set1} action={action}/>
          );
          expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
          expect(wrapper.find(ListItem).length).toBe(1);
          wrapper.setProps({set:{id:1}});
          wrapper.update();
          wrapper.render();
          expect(wrapper.find(ListItem).length).toBe(3);
      });
     
      test("handleClose prop", () => {
        let show = true;
        let set1={id:1};
        const action = jest.fn();
        const wrapper = mount(
            <AddToSetDialog show={show} set={set1} action={action}/>
          );
          act(() => {
            wrapper
              .find(Dialog)
              .at(0)
              .props()
              .onClose();
          });
          wrapper.render();
          wrapper.update();
          expect(action).toHaveBeenCalledTimes(1);
          expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
          expect(wrapper.find(ListItem).length).toBe(3);
      });

      test("action(FOLDER, true) prop", () => {
        let show = true;
        let set1={id:1};
        const action = jest.fn();
        const wrapper = mount(
            <AddToSetDialog show={show} set={set1} action={action}/>
          );
          act(() => {
            wrapper
              .find(ListItem)
              .at(0)
              .props()
              .onClick();
          });
          wrapper.render();
          wrapper.update();
          expect(action).toHaveBeenCalledTimes(1);
          expect(action.mock.calls[0][0]).toBe(FOLDER);
          expect(action.mock.calls[0][1]).toBeTruthy();
          expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
          expect(wrapper.find(ListItem).length).toBe(3);
      });
      
      test("action(FILE) prop", () => {
        let show = true;
        let set1={id:1};
        const action = jest.fn();
        const wrapper = mount(
            <AddToSetDialog show={show} set={set1} action={action}/>
          );
          act(() => {
            wrapper
              .find(ListItem)
              .at(2)
              .props()
              .onClick();
          });
          wrapper.render();
          wrapper.update();
          expect(action).toHaveBeenCalledTimes(1);
          expect(action.mock.calls[0][0]).toBe(FILE);
          expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
          expect(wrapper.find(ListItem).length).toBe(3);
      });

}); 