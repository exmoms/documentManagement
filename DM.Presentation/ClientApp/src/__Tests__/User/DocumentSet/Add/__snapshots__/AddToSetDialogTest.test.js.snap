// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToSetDialog AddToSetDialog matches snapshot 1`] = `"<div role=\\"presentation\\" class=\\"MuiDialog-root\\" style=\\"position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;\\"><div class=\\"MuiBackdrop-root\\" aria-hidden=\\"true\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div><div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div><div class=\\"MuiDialog-container MuiDialog-scrollPaper\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"none presentation\\" tabindex=\\"-1\\"><div class=\\"MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXs MuiPaper-elevation24 MuiPaper-rounded\\" role=\\"dialog\\" aria-labelledby=\\"simple-dialog-title\\"><div style=\\"text-align: left;\\"><div class=\\"MuiDialogTitle-root\\" id=\\"simple-dialog-title\\"><h2 class=\\"MuiTypography-root MuiTypography-h6\\">actions</h2></div></div><ul class=\\"MuiList-root MuiList-padding\\"><div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button Mui-focusVisible Mui-focusVisible\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\"><div class=\\"MuiListItemAvatar-root\\"><div class=\\"MuiAvatar-root MuiAvatar-circle makeStyles-avatar-1 MuiAvatar-colorDefault\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z\\"></path></svg></div></div><div class=\\"MuiListItemText-root\\"><span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock\\">documentset_addnewdocumentset</span></div><span class=\\"MuiTouchRipple-root\\"></span></div><div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button Mui-focusVisible Mui-focusVisible\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\"><div class=\\"MuiListItemAvatar-root\\"><div class=\\"MuiAvatar-root MuiAvatar-circle makeStyles-avatar-1 MuiAvatar-colorDefault\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z\\"></path></svg></div></div><div class=\\"MuiListItemText-root\\"><span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock\\">add_existing_set</span></div><span class=\\"MuiTouchRipple-root\\"></span></div><div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button Mui-focusVisible Mui-focusVisible\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\"><div class=\\"MuiListItemAvatar-root\\"><div class=\\"MuiAvatar-root MuiAvatar-circle makeStyles-avatar-1 MuiAvatar-colorDefault\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z\\"></path></svg></div></div><div class=\\"MuiListItemText-root\\"><span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock\\">add_document_to_set</span></div><span class=\\"MuiTouchRipple-root\\"></span></div></ul></div></div><div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div></div>"`;
